File: output/GPT4o-Mini/xCodeEval_4000/C++/Python/c985a9eab69e1d2c5b6b0a4a14eaf7be.py
import sys
import random
from collections import defaultdict

class Item:
    def __init__(self, key):
        self.key = key
        self.prior = random.randint(0, 2**31 - 1)
        self.cnt = 1
        self.l = None
        self.r = None

def update(t):
    if t is None:
        return
    t.cnt = (0 if t.l is None else t.l.cnt) + (0 if t.r is None else t.r.cnt) + 1

def split(t, key):
    if t is None:
        return None, None
    elif key < t.key or (key == t.key and random.randint(0, 1)):
        l, t.l = split(t.l, key)
        r = t
    else:
        t.r, r = split(t.r, key)
        l = t
    update(l)
    update(r)
    return l, r

def wsplit(t, key):
    if t is None:
        return None, None
    elif key < t.key or key == t.key:
        l, t.l = wsplit(t.l, key)
        r = t
    else:
        t.r, r = wsplit(t.r, key)
        l = t
    update(l)
    update(r)
    return l, r

def insert(t, it):
    if t is None:
        t = it
    elif it.prior > t.prior:
        it.l, it.r = split(t, it.key)
        t = it
    else:
        insert(t.l if it.key < t.key or (it.key == t.key and random.randint(0, 1)) else t.r, it)
    update(t)

def merge(t, l, r):
    if l is None or r is None:
        t = l if l else r
    elif l.prior > r.prior:
        merge(l.r, l.r, r)
        t = l
    else:
        merge(r.l, l, r.l)
        t = r
    update(t)

def erase(t, key):
    if t.key == key:
        merge(t, t.l, t.r)
    else:
        erase(t.l if key < t.key else t.r, key)
    update(t)

def ans(a):
    t = None
    total = 0
    for i in a:
        l, r = wsplit(t, i[1])
        total += 0 if r is None else r.cnt
        merge(t, l, r)
        insert(t, Item(i[1]))
    return total

def main():
    input = sys.stdin.read
    data = input().splitlines()
    n, w = map(int, data[0].split())
    b = []
    for i in range(1, n + 1):
        a, v = map(int, data[i].split())
        b.append(((abs(a), abs(v - w)), (abs(a), abs(v + w))))
    
    b.sort(key=lambda x: (x[0], -x[1][1]))
    print(ans(b))

if __name__ == "__main__":
    main()

File: output/GPT4o-Mini/xCodeEval_4000/C++/Go/c985a9eab69e1d2c5b6b0a4a14eaf7be.go
package main

import (
	"fmt"
	"math/rand"
	"sort"
)

type Item struct {
	key   [2]int
	prior int
	cnt   int
	l, r  *Item
}

func NewItem(key [2]int) *Item {
	return &Item{key: key, prior: rand.Int(), cnt: 1, l: nil, r: nil}
}

func (t *Item) Update() {
	if t == nil {
		return
	}
	t.cnt = 1
	if t.l != nil {
		t.cnt += t.l.cnt
	}
	if t.r != nil {
		t.cnt += t.r.cnt
	}
}

func Split(t *Item, key [2]int) (*Item, *Item) {
	if t == nil {
		return nil, nil
	}
	if less(key, t.key) || (equal(key, t.key) && rand.Int()%2 == 0) {
		l, r := Split(t.l, key)
		t.l = r
		t.Update()
		return l, t
	}
	l, r := Split(t.r, key)
	t.r = l
	t.Update()
	return t, r
}

func WSplit(t *Item, key [2]int) (*Item, *Item) {
	if t == nil {
		return nil, nil
	}
	if less(key, t.key) || equal(key, t.key) {
		l, r := WSplit(t.l, key)
		t.l = r
		t.Update()
		return l, t
	}
	l, r := WSplit(t.r, key)
	t.r = l
	t.Update()
	return t, r
}

func Insert(t **Item, it *Item) {
	if *t == nil {
		*t = it
	} else if it.prior > (*t).prior {
		l, r := Split(*t, it.key)
		*t = it
		(*t).l = l
		(*t).r = r
	} else {
		if less(it.key, (*t).key) || (equal(it.key, (*t).key) && rand.Int()%2 == 0) {
			Insert(&(*t).l, it)
		} else {
			Insert(&(*t).r, it)
		}
	}
	(*t).Update()
}

func Merge(t **Item, l, r *Item) {
	if l == nil || r == nil {
		*t = l
		if l == nil {
			*t = r
		}
		return
	}
	if l.prior > r.prior {
		Merge(&l.r, l.r, r)
		*t = l
	} else {
		Merge(&r.l, l, r.l)
		*t = r
	}
	(*t).Update()
}

func Erase(t **Item, key [2]int) {
	if (*t).key == key {
		Merge(t, (*t).l, (*t).r)
	} else {
		if less(key, (*t).key) {
			Erase(&(*t).l, key)
		} else {
			Erase(&(*t).r, key)
		}
	}
	if *t != nil {
		(*t).Update()
	}
}

func less(a, b [2]int) bool {
	return int64(a[0])*int64(b[1]) < int64(a[1])*int64(b[0])
}

func equal(a, b [2]int) bool {
	return int64(a[0])*int64(b[1]) == int64(a[1])*int64(b[0])
}

func Ans(a [][2][2]int) int64 {
	var t *Item
	var ans int64
	for _, i := range a {
		l, r := WSplit(t, i[1])
		if r != nil {
			ans += int64(r.cnt)
		}
		Merge(&t, l, r)
		Insert(&t, NewItem(i[1]))
	}
	return ans
}

func main() {
	var n, w int
	fmt.Scan(&n, &w)
	b := make([][2][2]int, n)
	for i := 0; i < n; i++ {
		var a, v int
		fmt.Scan(&a, &v)
		b[i] = [2][2]int{{abs(a), abs(v - w)}, {abs(a), abs(v + w)}}
	}
	sort.Slice(b, func(i, j int) bool {
		if b[i][0] != b[j][0] {
			return b[i][0] < b[j][0]
		}
		return b[i][1] > b[j][1]
	})
	fmt.Println(Ans(b))
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}
